redis/src/cluster_async/mod.rs

```diff
@@ -1145,22 +1217,92 @@ where
         }
     }
 
+    // Validate all existing user connections and try to reconnect if nessesary.
+    // In addition, as a safety measure, drop nodes that do not have any assigned slots.
+    // This function serves as a cheap alternative to slot_refresh() and thus can be used much more frequently.
+    // The function does not discover the topology from the cluster and assumes the cached topology is valid.
+    // In addition, the validation is done by peeking at the state of the underlying transport w/o overhead of additional commands to server.
+    async fn validate_all_user_connections(inner: Arc<InnerCore<C>>) {
+        let mut all_valid_conns = HashMap::new();
+        let mut all_nodes_with_slots = HashSet::new();
+        // prep connections and clean out these w/o assigned slots, as we might have established connections to unwanted hosts
+        {
+            let mut nodes_to_delete = Vec::new();
+            let connections_container = inner.conn_lock.read().await;
+
+            connections_container
+                .slot_map
+                .addresses_for_all_nodes()
+                .iter()
+                .for_each(|addr| {
+                    all_nodes_with_slots.insert(String::from(*addr));
+                });
+
+            connections_container
+                .all_node_connections()
+                .for_each(|(addr, con)| {
+                    if all_nodes_with_slots.contains(&addr) {
+                        all_valid_conns.insert(addr.clone(), con.clone());
+                    } else {
+                        nodes_to_delete.push(addr.clone());
+                    }
+                });
+
+            for addr in nodes_to_delete.iter() {
+                connections_container.remove_node(addr);
+            }
+        }
+
+        // identify nodes with closed connection
+        let mut addrs_to_refresh = Vec::new();
+        for addr_and_fut in all_valid_conns.iter() {
```

Comment:
Use the below loop to make the code more readable (better than using `.0` and `.1`
```rust
for (addr, fut) in &all_valid_conns {
...
}
```